File: frontend/src/model/Formal.ts
import React from "react";
import { BillDto } from "../helpers/parseBill";
import { Group } from "./Group";
import { ManualTicket } from "./ManualTicket";
import { AdminTicket, Ticket } from "./Ticket";

export interface Formal {
    id: string;
    name: string;
    menu: string;
    price: number;
    guestPrice: number;
    guestLimit: number;
    firstSaleTickets: number;
    ticketsRemaining: number;
    firstSaleGuestTickets: number;
    guestTicketsRemaining: number;
    firstSaleStart: Date;
    secondSaleStart: Date;
    secondSaleTickets: number;
    secondSaleGuestTickets: number;
    saleEnd: Date;
    dateTime: Date;
    hasGuestList: boolean;
    isVisible: boolean;
    // guestList: 
    // hidden?
    myTickets?: Ticket[];
    billId?: string;
    bill?: BillDto;
    groups?: Group[];
    ticketSales?: AdminTicket[];
    manualTickets?: ManualTicket[];
    queueLength?: number;
}

export const FormalContext = React.createContext<Formal>({} as Formal);

File: backend/model/formal.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Formal struct {
	Model
	Name                   string         `json:"name"`
	Menu                   string         `json:"menu"`
	Price                  float32        `json:"price" gorm:"type:decimal(10,2)"`
	GuestPrice             float32        `json:"guestPrice" gorm:"type:decimal(10,2)"`
	GuestLimit             uint           `json:"guestLimit"`
	FirstSaleTickets       uint           `json:"firstSaleTickets"`
	FirstSaleGuestTickets  uint           `json:"firstSaleGuestTickets"`
	FirstSaleStart         time.Time      `json:"firstSaleStart"`
	SecondSaleTickets      uint           `json:"secondSaleTickets"`
	SecondSaleGuestTickets uint           `json:"secondSaleGuestTickets"`
	SecondSaleStart        time.Time      `json:"secondSaleStart"`
	SaleEnd                time.Time      `json:"saleEnd"`
	DateTime               time.Time      `json:"dateTime"`
	HasGuestList           bool           `json:"hasGuestList"`
	IsVisible              bool           `json:"isVisible"`
	BillID                 *uuid.UUID     `json:"billId,omitempty"`
	Bill                   *Bill          `json:"bill,omitempty"`
	TicketSales            []Ticket       `json:"-"`
	ManualTickets          []ManualTicket `json:"-"`
	Groups                 []Group        `json:"groups,omitempty" gorm:"many2many:formal_groups;"`
	// TODO: boolean for public guestlist
	// Use pointers for updates?
}
